clients/drcachesim/tests/scheduler_unit_tests.cpp

```diff
@@ -6038,6 +6038,269 @@ test_kernel_switch_sequences()
+            inputs.push_back(make_exit(tid));
+            readers.emplace_back(
+                std::unique_ptr<mock_reader_t>(new mock_reader_t(inputs)),
+                std::unique_ptr<mock_reader_t>(new mock_reader_t()), tid);
+        }
+        sched_inputs.emplace_back(std::move(readers));
+        scheduler_t::scheduler_options_t sched_ops(scheduler_t::MAP_TO_ANY_OUTPUT,
+                                                   scheduler_t::DEPENDENCY_TIMESTAMPS,
+                                                   scheduler_t::SCHEDULER_DEFAULTS,
+                                                   /*verbosity=*/3);
+        sched_ops.quantum_duration_instrs = INSTR_QUANTUM;
+        sched_ops.kernel_syscall_reader = std::move(syscall_reader);
+        sched_ops.kernel_syscall_reader_end = std::move(syscall_reader_end);
+        scheduler_t scheduler;
+        if (scheduler.init(sched_inputs, NUM_OUTPUTS, std::move(sched_ops)) !=
+            scheduler_t::STATUS_SUCCESS)
+            assert(false);
+
+        // We have a custom version of run_lockstep_simulation here for more precise
+        // testing of the markers and instructions and interfaces.
+        // We record the entire sequence for a detailed check of some records, along with
+        // a character representation for a higher-level view of the whole sequence.
+        std::vector<scheduler_t::stream_t *> outputs(NUM_OUTPUTS, nullptr);
+        std::vector<bool> eof(NUM_OUTPUTS, false);
+        for (int i = 0; i < NUM_OUTPUTS; i++)
+            outputs[i] = scheduler.get_stream(i);
+        int num_eof = 0;
+        std::vector<std::vector<memref_t>> refs(NUM_OUTPUTS);
+        std::vector<std::string> sched_as_string(NUM_OUTPUTS);
+        std::vector<memref_tid_t> prev_tid(NUM_OUTPUTS, INVALID_THREAD_ID);
+        std::vector<bool> in_syscall(NUM_OUTPUTS, false);
+        std::vector<uint64> prev_in_ord(NUM_OUTPUTS, 0);
+        std::vector<uint64> prev_out_ord(NUM_OUTPUTS, 0);
+        while (num_eof < NUM_OUTPUTS) {
+            for (int i = 0; i < NUM_OUTPUTS; i++) {
+                if (eof[i])
+                    continue;
+                memref_t memref;
+                scheduler_t::stream_status_t status = outputs[i]->next_record(memref);
+                if (status == scheduler_t::STATUS_EOF) {
+                    ++num_eof;
+                    eof[i] = true;
+                    continue;
+                }
+                if (status == scheduler_t::STATUS_IDLE) {
+                    sched_as_string[i] += '_';
+                    continue;
+                }
+                assert(status == scheduler_t::STATUS_OK);
+                refs[i].push_back(memref);
+                if (memref.instr.tid != prev_tid[i]) {
+                    if (!sched_as_string[i].empty())
+                        sched_as_string[i] += ',';
+                    sched_as_string[i] +=
+                        'A' + static_cast<char>(memref.instr.tid - TID_BASE);
+                }
```

Comment:
Lines 6118-here plus some below look identical to test_kernel_switch_sequences() -- could we share this code via a new helper that lets each test customize the middle of the loop?

Reply:
Done.